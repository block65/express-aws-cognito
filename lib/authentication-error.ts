import { CustomError, Status } from '@block65/custom-error';

export class AuthenticationError extends CustomError {
  public constructor(message: string, err?: Error) {
    super(message, err);
    this.setName('AuthenticationError');
    this.code = Status.PERMISSION_DENIED;
  }
}
