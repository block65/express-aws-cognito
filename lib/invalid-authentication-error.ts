import { CustomError, Status } from '@block65/custom-error';

export class InvalidAuthenticationError extends CustomError {
  public code = Status.INVALID_ARGUMENT;

  constructor(message?: string, cause?: Error) {
    super(message || 'Invalid Authentication', cause);
  }
}
