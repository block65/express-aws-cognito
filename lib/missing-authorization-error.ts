import { CustomError, Status } from '@block65/custom-error';

export class MissingAuthorizationError extends CustomError {
  public constructor(message: string, err?: Error) {
    super(message, err);
    this.setName('MissingAuthorizationError');
    this.statusCode = Status.UNAUTHENTICATED;
  }
}
