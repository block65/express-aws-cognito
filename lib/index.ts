import type { RequestHandler } from 'express';
import { expressAsyncWrap } from '@block65/express-async-wrapper';
import {
  AwsCognitoAuthOptions,
  awsCognitoTokenVerifierFactory,
} from '@block65/aws-cognito-auth';
import { MissingAuthorizationError } from './missing-authorization-error';

export { MissingAuthorizationError };

export function expressAwsCognito({
  region,
  userPoolId,
  userIdGenerator,
}: AwsCognitoAuthOptions): RequestHandler {
  if (!region) {
    throw new Error('Missing/undefined issuer argument');
  }

  const verifier = awsCognitoTokenVerifierFactory({
    region,
    userPoolId,
    userIdGenerator,
  });

  return expressAsyncWrap(
    async (req, res, next): Promise<void> => {
      const [scheme = '', jwt = ''] =
        req.header('authorization')?.split(' ') || [];

      if (scheme.toLowerCase() !== 'bearer') {
        throw new MissingAuthorizationError(
          `Expected Authorization method: Bearer - saw ${scheme}`,
        ).debug({ scheme, jwt });
      }

      if (!jwt) {
        throw new MissingAuthorizationError(
          'Invalid or Missing Bearer token',
        ).debug({
          scheme,
          jwt,
        });
      }

      res.locals.token = await verifier(jwt);

      next();
    },
  );
}
