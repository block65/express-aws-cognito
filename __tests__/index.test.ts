import type { NextFunction, Request, Response } from 'express';
import express from 'express';
import {
  TokenInvalidError,
  TokenUnsuitableError,
} from '@block65/aws-cognito-auth';
import { jest, describe, test } from '@jest/globals';
// eslint-disable-next-line import/extensions
import { expressAwsCognito } from '../lib/index';
// eslint-disable-next-line import/extensions
import { MissingAuthorizationError } from '../lib/missing-authorization-error';

function mockReq(options: { headers?: Record<string, string> } = {}): Request {
  return {
    body: {},
    cookies: {},
    query: {},
    params: {},
    headers: options.headers || {},
    method: 'get',
    url: '/',
    listeners: () => [],
    get: jest.fn(),
    resume: jest.fn().mockReturnThis(),
    ...options,
  } as unknown as Request;
}

function mockRes(): Response {
  return {
    setHeader: jest.fn().mockReturnThis(),
    status: jest.fn().mockReturnThis(),
    send: jest.fn().mockReturnThis(),
    end: jest.fn().mockReturnThis(),
  } as unknown as Response;
}

function testApp(
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<unknown> {
  const app = express();
  app.use(expressAwsCognito({ region: 'local', userPoolId: 'issuer' }));

  return new Promise<void>((resolve) => {
    app(req, res, (err: unknown) => {
      next(err);
      resolve();
    });
  });
}

describe('express', () => {
  test('should throw MissingAuthorizationError with invalid headers', async () => {
    await testApp(mockReq(), mockRes(), (err: any) => {
      expect(err).toBeInstanceOf(MissingAuthorizationError);
    });
    expect.assertions(1);
  });

  test('should throw MissingAuthorizationError with no headers', async () => {
    await testApp(
      mockReq({
        headers: {
          authorization: 'OMG DED',
        },
      }),
      mockRes(),
      (err: any) => {
        expect(err).toBeInstanceOf(MissingAuthorizationError);
        expect(err.message).toContain('Invalid Authorization scheme');
      },
    );
    expect.assertions(2);
  });

  test('should throw MissingAuthorizationError with missing JWT ', async () => {
    await testApp(
      mockReq({
        headers: {
          authorization: 'Bearer',
        },
      }),
      mockRes(),
      (err: any) => {
        expect(err).toBeInstanceOf(MissingAuthorizationError);
        expect(err.message).toContain('Invalid');
      },
    );
    expect.assertions(2);
  });

  test('should throw TokenError with bad JWT ', async () => {
    await testApp(
      mockReq({
        headers: {
          authorization: 'Bearer DED',
        },
      }),
      mockRes(),
      (err: any) => {
        expect(err).toBeInstanceOf(TokenInvalidError);
        expect(err.message).toContain('parse');
      },
    );
    expect.assertions(2);
  });

  test('should throw TokenError with fake JWT ', async () => {
    await testApp(
      mockReq({
        headers: {
          authorization:
            'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c',
        },
      }),
      mockRes(),
      (err: any) => {
        expect(err).toBeInstanceOf(TokenUnsuitableError);
        // expect(err.debug()).toMatchObject({
        //   code: 'invalid_token',
        //   inner: expect.any(Error),
        // });
        // expect(err.message).toContain('Invalid');
      },
    );
    expect.assertions(1);
  });
});
