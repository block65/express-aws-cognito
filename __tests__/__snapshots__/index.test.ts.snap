// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`express should throw TokenError with bad JWT  1`] = `
Object {
  "code": 3,
  "debug": Object {
    "decoded": null,
  },
  "httpStatusCode": 400,
  "message": "Unparseable token",
  "name": "Error",
  "stack": Any<String>,
  "status": "INVALID_ARGUMENT",
}
`;

exports[`express should throw TokenError with fake JWT  1`] = `
Object {
  "code": 3,
  "debug": Object {
    "decoded": Object {
      "header": Object {
        "alg": "HS256",
        "typ": "JWT",
      },
      "payload": Object {
        "iat": 1516239022,
        "name": "John Doe",
        "sub": "1234567890",
      },
      "signature": "SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c",
    },
  },
  "httpStatusCode": 400,
  "message": "Missing key id",
  "name": "Error",
  "stack": Any<String>,
  "status": "INVALID_ARGUMENT",
}
`;

exports[`express should throw correct error with invalid headers 1`] = `
Object {
  "code": 3,
  "debug": Object {
    "headerValue": Array [
      "Boring",
      "JWT",
    ],
    "scheme": "Boring",
  },
  "details": Array [
    Object {
      "violations": Array [
        Object {
          "description": "Invalid scheme",
          "field": "authorization",
        },
      ],
    },
  ],
  "httpStatusCode": 400,
  "message": "Invalid Authentication",
  "name": "Error",
  "stack": Any<String>,
  "status": "INVALID_ARGUMENT",
}
`;

exports[`express should throw correct error with missing JWT  1`] = `
Object {
  "code": 3,
  "debug": Object {
    "headerValue": Array [
      "Bearer",
    ],
    "jwt": "",
  },
  "details": Array [
    Object {
      "violations": Array [
        Object {
          "description": "Missing bearer token",
          "field": "authorization",
        },
      ],
    },
  ],
  "httpStatusCode": 400,
  "message": "Invalid Authentication",
  "name": "Error",
  "stack": Any<String>,
  "status": "INVALID_ARGUMENT",
}
`;
